CRM/Contribute/BAO/ContributionRecur.php

```diff
@@ -962,4 +955,41 @@ public static function getInactiveStatuses() {
     return self::$inactiveStatuses;
   }
 
+  /**
+   * Get options for the called BAO object's field.
+   *
+   * This function can be overridden by each BAO to add more logic related to context.
+   * The overriding function will generally call the lower-level CRM_Core_PseudoConstant::get
+   *
+   * @param string $fieldName
+   * @param string $context
+   * @see CRM_Core_DAO::buildOptionsContext
+   * @param array $props
+   *   whatever is known about this bao object.
+   *
+   * @return array|bool
+   */
+  public static function buildOptions($fieldName, $context = NULL, $props = array()) {
+
+    switch ($fieldName) {
```

Comment:
I feel like we are duplicating some logic in \Civi\Payment\System - we should rationalise

Reply:
In this case we just want a list of all the payment processor names without doing any initialisation etc.  Whereas \Civi\Payment\System is initialising and returning specific processors with the intent of actually using that processor directly.  Agree that further cleanup would be good but I don't see anywhere else in CiviCRM which actually does what I'm putting in `buildOptions()` (except the bit of code I'm moving from the contribution search form `recurringContribution()`).